File: apps/ios/Sources/CoreDomain/Observations/ObservationSymptom.swift
import Foundation

/// What an observer SAW, never what caused it: `leafSpots` is visible,
/// `blight` would be a diagnosis they are guessing at.
///
/// Deliberately not `ImageAnalysisKind`. That enum is a model's vocabulary for
/// what it proposed; this one is a person's for what they reported, and the two
/// are never merged — see `ObservationSymptom`.
///
/// Source: packages/api-contracts/openapi.yaml, `ObservationSymptomKind`.
public enum ObservationSymptomKind: String, Codable, Equatable, Sendable, CaseIterable {
    case leafSpots = "leaf_spots"
    case leafYellowing = "leaf_yellowing"
    case leafCurling = "leaf_curling"
    case wilting
    case holesOrChewing = "holes_or_chewing"
    case mouldOrMildew = "mould_or_mildew"
    case dieback
    case stuntedGrowth = "stunted_growth"
    case unusualGrowth = "unusual_growth"
}

/// Three values rather than a numeric scale: two people would not report the
/// same leaf as the same number out of ten.
public enum ObservationSymptomSeverity: String, Codable, Equatable, Sendable, CaseIterable {
    case mild
    case moderate
    case severe
}

/// A symptom a PERSON reported on an observation (P11-MEDIA-01).
///
/// Never an `ImageAnalysisResult`, which is what a model proposed and what a
/// reviewer decided about it. The two carry different weight in a health
/// review and share no storage, no vocabulary, and no reference; a screen that
/// showed them as one list would present a guess as testimony.
///
/// At most one per kind per observation: seeing the same symptom worse next
/// week is a new observation.
public struct ObservationSymptom: Equatable, Sendable, Identifiable {
    public let id: String
    public let kind: ObservationSymptomKind
    public let severity: ObservationSymptomSeverity
    public let createdAt: Date

    public init(
        id: String,
        kind: ObservationSymptomKind,
        severity: ObservationSymptomSeverity,
        createdAt: Date
    ) {
        self.id = id
        self.kind = kind
        self.severity = severity
        self.createdAt = createdAt
    }
}

/// A symptom being reported, before the server assigns it an id.
public struct ObservationSymptomInput: Equatable, Sendable, Identifiable {
    public var kind: ObservationSymptomKind
    public var severity: ObservationSymptomSeverity

    /// The kind: unique across an observation's symptoms by construction.
    public var id: ObservationSymptomKind { kind }

    public init(kind: ObservationSymptomKind, severity: ObservationSymptomSeverity) {
        self.kind = kind
        self.severity = severity
    }
}

File: services/api/src/modules/observations-history/domain/observation-symptom.ts
/**
 * A symptom a PERSON reported on an observation (P11-MEDIA-01).
 *
 * Deliberately not the same thing as an `image_analysis_result`, and
 * deliberately not stored with one. That row is a model's proposal and a
 * reviewer's disposition; this one is testimony. Reading either as the other
 * would misstate how much weight a health review should give it, so the two
 * share no table, no foreign key, and no vocabulary — see this table's own
 * migration header for why a single table with an `authored_by` discriminator
 * was rejected.
 *
 * `kind` describes what was VISIBLE, never what caused it: `leaf_spots` is
 * something a gardener can see, `blight` is a diagnosis they would be guessing
 * at. The closed set exists for the same reason
 * `observation_measurement.kind`'s does — the value arrives in a request body,
 * and an open column would accept a symptom nothing can query.
 *
 * At most one row per (observation, kind), matching
 * `observation_symptom_unique_kind`: seeing the same symptom worse next week
 * is a new observation.
 *
 * Source: migrations/1788400000000_observation-symptoms.sql;
 * architecture/plant-intelligence-and-visual-journal.md, section 8.1.
 */

import { SharedErrorCode } from '@verdery/api-contracts';
import { ValidationError } from '../../../platform/errors/application-error.js';
import type { Uuid } from '../../../shared/identifiers/uuid.js';

export type ObservationSymptomKind =
  | 'leaf_spots'
  | 'leaf_yellowing'
  | 'leaf_curling'
  | 'wilting'
  | 'holes_or_chewing'
  | 'mould_or_mildew'
  | 'dieback'
  | 'stunted_growth'
  | 'unusual_growth';

export const OBSERVATION_SYMPTOM_KINDS: readonly ObservationSymptomKind[] = [
  'leaf_spots',
  'leaf_yellowing',
  'leaf_curling',
  'wilting',
  'holes_or_chewing',
  'mould_or_mildew',
  'dieback',
  'stunted_growth',
  'unusual_growth',
];

/** Three values, not a numeric scale: two people would not report the same leaf as the same number out of ten. */
export type ObservationSymptomSeverity = 'mild' | 'moderate' | 'severe';

export const OBSERVATION_SYMPTOM_SEVERITIES: readonly ObservationSymptomSeverity[] = [
  'mild',
  'moderate',
  'severe',
];

export interface ObservationSymptom {
  readonly id: Uuid;
  readonly observationId: Uuid;
  readonly kind: ObservationSymptomKind;
  readonly severity: ObservationSymptomSeverity;
  readonly createdAt: Date;
}

function invalidField(code: string, pointer: string, message: string): ValidationError {
  return new ValidationError(SharedErrorCode.RequestInvalid, message, {
    details: [{ code, pointer }],
  });
}

export function createObservationSymptom(
  id: Uuid,
  observationId: Uuid,
  rawKind: string,
  rawSeverity: string,
  now: Date,
): ObservationSymptom {
  if (!OBSERVATION_SYMPTOM_KINDS.includes(rawKind as ObservationSymptomKind)) {
    throw invalidField(
      'observation_symptom.kind.invalid',
      '/kind',
      `kind must be one of: ${OBSERVATION_SYMPTOM_KINDS.join(', ')}.`,
    );
  }
  if (!OBSERVATION_SYMPTOM_SEVERITIES.includes(rawSeverity as ObservationSymptomSeverity)) {
    throw invalidField(
      'observation_symptom.severity.invalid',
      '/severity',
      `severity must be one of: ${OBSERVATION_SYMPTOM_SEVERITIES.join(', ')}.`,
    );
  }

  return {
    id,
    observationId,
    kind: rawKind as ObservationSymptomKind,
    severity: rawSeverity as ObservationSymptomSeverity,
    createdAt: now,
  };
}
